lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,1039 @@
+		"ID %d", requestID)
+
+	err = s.stream.Send(&walletrpc.SignCoordinatorRequest{
+		RequestId:       requestID,
+		SignRequestType: req,
+	})
+	if err != nil {
+		return false, err
+	}
+
+	newTimeout := timeout - time.Since(startTime)
+
+	if time.Since(startTime) > timeout {
+		return false, ErrRequestTimeout
+	}
+
+	resp, err := s.getResponse(requestID, newTimeout)
+	if err != nil {
+		return false, err
+	}
+
+	signResp := resp.GetPong()
+	if !signResp {
+		return false, ErrUnexpectedResponse
+	}
+
+	return signResp, nil
+}
+
+// DeriveSharedKey sends a SharedKeyRequest to the remote signer and waits for
+// the corresponding response.
+//
+// NOTE: This is part of the RemoteSignerRequests interface.
+func (s *SignCoordinator) DeriveSharedKey(_ context.Context,
+	in *signrpc.SharedKeyRequest,
+	_ ...grpc.CallOption) (*signrpc.SharedKeyResponse, error) {
+
+	done, err := s.registerRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	defer done()
+
+	err = s.WaitUntilConnected()
+	if err != nil {
+		return nil, err
+	}
+
+	requestID := s.getNextRequestID()
+
+	req := &walletrpc.SignCoordinatorRequest_SharedKeyRequest{
+		SharedKeyRequest: in,
+	}
+
+	cleanUp := s.createResponseChannel(requestID)
+	defer cleanUp()
+
+	log.Debugf("Sending a signrpc.SharedKeyRequest to the remote signer "+
+		"with request ID %d", requestID)
+
+	err = s.stream.Send(&walletrpc.SignCoordinatorRequest{
+		RequestId:       requestID,
+		SignRequestType: req,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := s.getResponse(requestID, s.requestTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	rpcResp := resp.GetSharedKeyResponse()
+	if rpcResp == nil {
+		return nil, ErrUnexpectedResponse
+	}
+
+	return rpcResp, nil
+}
+
+// MuSig2Cleanup sends a MuSig2CleanupRequest to the remote signer and waits for
+// the corresponding response.
+//
+// NOTE: This is part of the RemoteSignerRequests interface.
+func (s *SignCoordinator) MuSig2Cleanup(_ context.Context,
+	in *signrpc.MuSig2CleanupRequest,
+	_ ...grpc.CallOption) (*signrpc.MuSig2CleanupResponse, error) {
+
+	done, err := s.registerRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	defer done()
+
+	err = s.WaitUntilConnected()
+	if err != nil {
+		return nil, err
+	}
+
+	requestID := s.getNextRequestID()
+
+	req := &walletrpc.SignCoordinatorRequest_MuSig2CleanupRequest{
+		MuSig2CleanupRequest: in,
+	}
+
+	cleanUp := s.createResponseChannel(requestID)
+	defer cleanUp()
+
+	log.Debugf("Sending a signrpc.MuSig2CleanupRequest to the remote "+
+		"signer with request ID %d", requestID)
+
+	err = s.stream.Send(&walletrpc.SignCoordinatorRequest{
+		RequestId:       requestID,
+		SignRequestType: req,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := s.getResponse(requestID, s.requestTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	rpcResp := resp.GetMuSig2CleanupResponse()
+	if rpcResp == nil {
+		return nil, ErrUnexpectedResponse
+	}
+
+	return rpcResp, nil
```

Comment:
I wonder if we could use Generics to cut down on the number of very similar looking lines of code... Could try to come up with an example, unless you want to try yourself first?

Reply:
Updated this to use a generic helper method :)! Hope that addresses your idea :rocket: